Auto-renew subscriptions

nospame opened the pull request:
## Product Description
Eligible subscriptions that are set to "auto renew" will automatically renew 30 or fewer days before their end date.

## Technical Summary
[Tech spec](https://docs.google.com/document/d/1kop7WTDE3HQdNhbWi0yoPbRjqdI4FLjFVHvtZ5CStJY/edit?usp=sharing), [ticket](https://dimagi.atlassian.net/browse/SAAS-18410)
Adds a task to automatically renew eligible subscriptions 30 or fewer days before (but not after) their end dates. Also creates new email templates to notify customers the subscription will be renewed, and creates+sends a prepayment invoice if applicable.

Base will be changed to `master` once auto-renewal/emails PR is merged.

## Feature Flag
`SHOW_AUTO_RENEWAL`

## Safety Assurance

### Safety story
Tested on staging that a subscription with auto-renewal enabled is correctly renewed with the new "Trigger Subscription Auto-Renewal" testing tool, that the notification email is sent to billing contacts, and a prepayment invoice is generated and sent when the new plan is a Pay Annually plan.

### Automated test coverage
Adds `TestAutoRenewableSubscriptions` to see that correct subscriptions are selected for auto-renewal, and `TestAutoRenewSubscription` to see that the correct actions are performed when auto renewing.

### QA Plan
This is the last major piece of auto-renewals before the feature gets QA.


### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

nospame committed:
Auto renew subscriptions 30 or fewer days from their end date

nospame committed:
Send subscription renewed email when auto renewed

nospame committed:
Generate prepayment invoice if auto renewing an annual plan

nospame committed:
Add TestAutoRenewableSubscriptions to see that correct subscriptions are selected for renewal

nospame committed:
Add TestAutoRenewSubscription

mjriley commented on corehq/apps/accounting/tasks.py:
```diff
@@ -421,6 +422,42 @@ def send_bookkeeper_email(month=None, year=None, emails=None):
+    subscription.renew_subscription(
+        adjustment_method=SubscriptionAdjustmentMethod.AUTO_RENEWAL,
+        new_version=new_plan_version,
+    )
```
Any idea if we need to think about a transaction here? I notice `renew_subscription` is already enclosed in a transaction, so I'm curious if we should be extending that to include fetching the current plan version. Is it an issue if the subscription plan's version were to change between when we initially fetch it and when we renew it?

nospame replied:
I hadn't considered it, but actually I think the whole thing should be a transaction. If for some reason we fail in notifying the customer, or in (kicking off the task to start) generating and sending an invoice, that should cause a rollback as well. https://github.com/dimagi/commcare-hq/pull/37094/commits/5d58f6777569ce02e66d1936caa00ee1f2c74d3d

mjriley commented on corehq/apps/accounting/tasks.py:
```diff
@@ -458,6 +460,27 @@ def auto_renew_subscription(subscription):
+        invoice_contacts = [
+            subscription.account.billingcontactinfo.email_list
+            if BillingContactInfo.objects.filter(account=next_subscription.account).exists() else None
+        ]
```
`email_list` looks like it would be an array, right? So this would set `invoice_contacts` to a 1-element or either the email list or `None`? Given the `not invoice_contacts` check right after, are the array brackets an error?
Nit, but it also seem more natural to use an empty array rather than `None` if no billing account exists.

nospame replied:
That's a good point. Resolved and moved to `get_contact_emails` on the prepayment invoice here https://github.com/dimagi/commcare-hq/pull/37094/commits/3db234d8b0932c2536d9119675e591ebbef98a4d

mjriley commented on corehq/apps/accounting/tests/test_tasks.py:
```diff
@@ -164,3 +164,78 @@ def test_no_subscription_ending_email_if_is_renewed(self):
+
+    def test_includes_subscription_less_than_30_days_left(self):
+        self._set_auto_renew_properties(
+            date_end=datetime.date.today() + datetime.timedelta(days=1)
```
I like our focus on boundary conditions here, but I see `days=1` and (essentially) `days=-1`. What about `days=0`. If we renew when `days=0`, I'd leave out `days=1`. If we don't, I'd leave out `days=-1`

mjriley reviewed (commented):
The `email_list` array is my main concern, assuming I'm understanding it correctly. Everything else looks good

nospame committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into ejp/auto-renewal/auto-renew-task

nospame committed:
Add TriggerAutoRenewal view and form for testing

nospame committed:
Move invoice contacts to get_contact_emails on the WirePrepaymentInvoice

Following pattern of other Invoice classes.

nospame committed:
Test at 0-day boundary

nospame committed:
isort imports

nospame committed:
Remove acks_late from auto_renew_subscriptions task

If the task dies one day and misses any subscriptions, they'll be picked up on the next day's run.

nospame commented on corehq/apps/accounting/tasks.py:
```diff
@@ -421,6 +422,42 @@ def send_bookkeeper_email(month=None, year=None, emails=None):
         })
 
 
+@periodic_task(run_every=crontab(minute=0, hour=0), acks_late=True)
```
I re-used this from the reminder email tasks, but those only send for an exact number of days left so it would be more of an issue if the task failed midway. This shouldn't need it since it will still try to renew the next day. https://github.com/dimagi/commcare-hq/pull/37094/commits/5a08a82483f76efdb377d4dbb2ec5f312f031f84

nospame committed:
Use transaction.atomic on auto_renew_subscription

mjriley commented on corehq/apps/accounting/views.py:
```diff
@@ -1310,6 +1312,32 @@ def post(self, request, *args, **kwargs):
+            messages.success(
+                request,
+                f'Successfully triggered subscription auto-renewal for "{domain}". '
+                'Any eligible subscriptions for this project will be automatically renewed.',
```
It looks like the domain can be empty, which would cause us to auto renew all subscriptions, regardless of domain. I assume that isn't an issue, because I'm guessing this is an internal tool? But the message would look a bit awkward if the domain was `""`. We might want to make that a conditional.

nospame replied:
I had a tough time finding exactly where it's happening, but the field is required in practice. It has the `required` html attr and does validate that it's not empty:
<img width="774" height="227" alt="image" src="https://github.com/user-attachments/assets/5ec6cda9-1442-4494-9557-5991932a4c94" />

mjriley reviewed (approved):
Left a few more comments I was unsure about, but trusting your judgment here. Approving so that, if none of the issues are real, it won't be blocked

nospame commented on corehq/apps/accounting/tasks.py:
```diff
@@ -451,6 +451,7 @@ def _get_auto_renewable_subscriptions(domain_name=None):
     return auto_renewable_subscriptions
 
 
+@transaction.atomic
```
I don't think it should be needed. The edition on a software plan should never change, and the software plan version on a subscription can't change. To get a new software plan on a domain's subscription, there would need to be a new subscription, at which point the current one would accurately give `is_renewed == True` because it uses a DB query to check if any other subscriptions are following it.

There might be a second or two gap between checking `not subscription.is_renewed` and the `auto_renew_subscription` function completing where someone could, _theoretically_, have already changed or renewed their plan causing a race condition. But I don't think this is realistically something that would happen.
